composer/datasets/in_context_learning_evaluation.py

```diff
@@ -85,582 +109,764 @@ def _make_padded_input(context_enc, continuation_enc, max_seq_len, pad_tok_id, p
 
-    def __len__(self):
-        return len(self.encoded_dataset)
+        Returns:
+            Dict: dictionary with the tokenized data
+        """
+        tokenized_example = {}
+        preamble = self.tokenizer(prompt_and_fewshot)
+        preamble['input_ids'] = self._fix_eos_on_preamble(preamble['input_ids'])
+        tokenized_example['preamble'] = preamble
+        if self.strip_data:
+            # rstrip context because a prompt ending in a space results in degenerate output
+            ctxt = ctxt.rstrip()
+        tokenized_example['context'] = self.tokenizer(ctxt, add_special_tokens=False)
+        return tokenized_example
+
+    def _prep_example(
+        self,
+        example: Dict,
+        example_idx: int,
+        num_fewshot: int,
+        prompt_string: str,
+        fewshot_rng: random.Random,
+    ) -> List[Dict[str, Any]]:
+        """
+        Prepares a single example from a HF Dataset into tokenized format with prompt and fewshot examples.
 
-    def collate_fn(self, data):
-        inputs, answers = [], []
-        cot_delimiter = ''
+        Each task consists of a context and a continuation as well as an optional prompt and optional list of
+        example context/continuation pairs which precede the test context/continuation pair.
 
-        for sample in data:
-            preamble, context, aliases = (sample['preamble'], sample['context'], sample['aliases'])
-            context_enc = preamble['input_ids'] + context['input_ids']
-            inp, _ = _make_padded_input(context_enc, [],
-                                        self.max_seq_len - self.max_answer_length,
-                                        self.pad_tok_id,
-                                        padding_side=self.padding_side)
+        Args:
+            example (Dict): A Dictionary from the hf dataset
+            example_idx (int): the index of example
+            num_fewshot (int): Number of examples context/continuation pairs to prepend to the test pair
+            prompt_string (str): The prompt to prepend to all inputs
+            fewshot_rng (random.Random): Random number generator to use for fewshot sampling
 
-            inputs.append(inp)
-            answers.append(aliases)
+        Returns:
+            Dict: contains a dictionary with the tokenized data
+        """
+        prompt_and_fewshot = self._generate_few_shot_text(num_fewshot, example_idx, prompt_string, fewshot_rng)
+        ctxt = self._construct_context(example, prompt_and_fewshot, add_answer=False)
+        tokenized_example = self._tokenize_example(prompt_and_fewshot, ctxt, example)
+        return tokenized_example
 
-            # We will search for the answer within the portion of the model response
-            # beginning with `cot_delimiter`
-            cot_delimiter = sample['cot_delimiter']
+    def collate_fn(self, data: Dict[str, Any]) -> Dict[str, Any]:
```

Comment:
How many distinct tasks follow this collate_fn? If it's just LM, then it might be better to leave collate_fn unimplemented in the base class and force the child classes to reimplement it

Reply:
I've updated the collate_fn considerably. New version is malleable but would like design feedback on if I actually improved usablitity.